app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/securities/SelectDutiesController.scala

```diff
@@ -16,18 +16,181 @@
+            .fold(emptyForm)(emptyForm.fill)
+          Ok(
+            selectDutiesPage(filledForm, securityId, dutiesAvailable, routes.SelectDutiesController.submit(securityId))
+          )
+        }.asFuture
+    )
+  }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Any"))
+  def submit(securityId: String): Action[AnyContent] = actionReadWriteJourney { implicit request => journey =>
+    processAvailableDuties[(SecuritiesJourney, Result)](
+      securityId: String,
+      journey: SecuritiesJourney,
+      request: Request[_],
+      error => {
+        logger.warn(s"No Available duties: $error")
+        (journey, Redirect(baseRoutes.IneligibleController.ineligible())).asFuture
+      },
+      dutiesAvailable => {
+        val form      = selectDutiesForm(dutiesAvailable.map(_._1))
+        val boundForm = form.bindFromRequest()
+        boundForm
+          .fold(
+            errors => {
+              logger.warn(
+                s"Selection of duties to be repaid failed for $securityId because of errors:" +
+                  s"${errors.mkString("", ",", "")}"
+              )
+              (
+                journey,
+                Ok(
+                  selectDutiesPage(
+                    boundForm,
+                    securityId,
+                    dutiesAvailable,
+                    routes.SelectDutiesController.submit(securityId)
+                  )
+                )
+              )
+            },
+            dutiesSelected => {
+              updateAndRedirect(journey, securityId, dutiesSelected)
+            }
+          )
+          .asFuture
+      }
+    )
+  }
+
+  private def updateAndRedirect(
+    journey: SecuritiesJourney,
+    securityId: String,
+    dutiesSelected: DutiesSelectedAnswer
+  ): (SecuritiesJourney, Result) =
+    journey
+      .selectAndReplaceTaxCodeSetForSelectedSecurityDepositId(securityId, dutiesSelected.map(_.taxCode).toList)
+      .fold(
+        error => {
+          logger.warn(error)
+          (journey, Redirect(controllers.routes.IneligibleController.ineligible()))
+        },
+        updatedJourney => {
+          //(updatedJourney, Redirect(routes.EnterClaimController.show(securityId, dutiesSelected.head.taxCode))) todo CDSR-1777
+          (
+            updatedJourney,
+            Redirect(
+              Call("GET", s"/claim-back-import-duty-vat/securities/enter-claim/$securityId/${dutiesSelected.head.taxCode}")
+            )
+          )}
+      )
+}
+object SelectDutiesController extends Logging {
+  val selectDutiesKey: String = "select-duties"
+
+  def getDescription(fullKey: String, messages: Messages): Option[String] =
+    if (messages.isDefinedAt(fullKey))
+      Some(messages(fullKey))
+    else {
+      logger.warn(s"no description found for $fullKey")
+      None
+    }
+
+  @SuppressWarnings(Array("org.wartremover.warts.Option2Iterable"))
+  def getTaxCodeDescriptions(dutiesAvailable: Seq[TaxCode], messages: Messages): List[(Duty, String)] = {
```

Comment:
I think message resolving should be part of the view layer logic, not on the controller level.

Reply:
In general I agree.  I thought about this and was 50/50 but then came down on the side of the controller for 2 reasons:
a) the collection needs to be ordered and there is no way to order it in the code (unlike in the backend which is not concerned with being multi lingual)  without either maintaining data separately from messages or accessing messages
b) the function is used in prod and test for verification and in this specific case only it seemed cleaner to put it in pure Scala.

Let me know if you prefer it changed